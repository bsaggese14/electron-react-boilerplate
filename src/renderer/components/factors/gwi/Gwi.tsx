import type { FactorName } from "../../../SupportedFactors";
import type { Props as GwiTableProps } from "../../../components/factors/gwi/GwiTable";

import { DateSelector } from "../../factors/gwi/DateSelector";
import { GwiTable } from "../../factors/gwi/GwiTable";

export type GwiFactor = GwiTableProps &
  FactorName & {
    effectiveDate: Date;
  };

// TO DO use effective date instead of hard coded date
export const Gwi = ({ config }: { config: GwiFactor }) => (
  <>
    <h4>General Wage Increase (GWI)</h4>
    <DateSelector defaultMonth={4} defaultDayOfWeek={6}></DateSelector>
    <GwiTable gwiObjects={config.gwiObjects}></GwiTable>
  </>
);
