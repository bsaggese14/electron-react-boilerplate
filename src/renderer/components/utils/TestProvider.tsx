import { Provider } from "jotai";
import { useHydrateAtoms } from "jotai/utils";

const HydrateAtoms = ({ initialValues, children }: { initialValues: any; children: any }) => {
  useHydrateAtoms(initialValues);
  return children;
};

export const TestProvider = ({ initialValues, children }: { initialValues: any; children: any }) => {
  return (
    <Provider>
      <HydrateAtoms initialValues={initialValues}>{children}</HydrateAtoms>
    </Provider>
  );
};
