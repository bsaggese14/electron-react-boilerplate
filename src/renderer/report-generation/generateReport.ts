import $ from "jquery";
import { obj, nestedObject, EmployeeData, calculateVacationHoursPerYear, roundNumber, gwiInputTableToJson, calculateWagesPaidForYear, getGwiDateFromDom, calculateProductiveHoursThisYear, calculatePostGwiWagesForEmployee, GwiData, getVacationTableData } from "./utils";
import { KeyMap } from "../components/data-table/employeeDataKeyMapAtom";

export const generateReport = ({ type, employeeDataArray, keyMap }: { type: string; employeeDataArray: EmployeeData[]; keyMap: KeyMap }) => {
  let gwiTable: GwiData = { percent: {}, bonus: {} };
  const vacationTableData = getVacationTableData(type);

  if (type === "current") {
    gwiTable = gwiInputTableToJson($(".current .gwiTable"));
  }
  if (type === "future") {
    gwiTable = gwiInputTableToJson($(".future .gwiTable"));
  }
  const gwiYears = Object.keys(gwiTable.percent);
  const gwiBonuses = gwiTable.bonus;

  let employeeWageData: nestedObject = {};
  let totalLaborCostPerYear: obj = {};
  let totalVacationHoursGrantedPerYear: obj = {};
  let totalProductiveHoursPerYear: obj = {};
  for (var employee in employeeDataArray) {
    const employeeData = employeeDataArray[employee];
    const employeeId = employeeData[keyMap.employeeId];
    const schedule = employeeData.Schedule;
    const startDate = new Date(employeeData[keyMap.startDate]);

    employeeWageData[employeeId] = {};
    const hourlyRate = employeeData[keyMap.hourlyWage] ?? "";
    employeeWageData[employeeId]["postGwiDayRatePerYear"] = calculatePostGwiWagesForEmployee({ hourlyRate, gwiTable });

    // calc pay per year
    let wagesTotalPerYear: obj = {};
    let wagesAndBonusPaidPerYear: obj = {};
    let vacationHoursPerYear: obj = {};
    let productiveHoursPerYear: obj = {};
    for (var i = 0; i < gwiYears.length; i++) {
      // if (gwiYears[i] != 2026) {
      //   continue;
      // }
      const currYear = parseInt(gwiYears[i]);

      const vacationHoursThisYear = roundNumber(calculateVacationHoursPerYear({ startDate, schedule, currYear, vacationTableData }), 2);
      vacationHoursPerYear[currYear] = vacationHoursThisYear;

      const holidays: Date[] = [];
      const productiveHoursThisYear = roundNumber(calculateProductiveHoursThisYear({ vacationHoursThisYear, holidays, schedule }), 2);
      productiveHoursPerYear[currYear] = productiveHoursThisYear;

      const gwiDate = getGwiDateFromDom(currYear, type);
      const totalPayment = calculateWagesPaidForYear({ employeeData, gwiDate, employeeWageData, currYear, keyMap });
      const roundedTotal = roundNumber(totalPayment, 2);
      wagesTotalPerYear[currYear] = roundedTotal;

      const bonusThisYear = gwiBonuses[currYear];
      wagesAndBonusPaidPerYear[currYear] = roundedTotal + bonusThisYear;

      totalLaborCostPerYear[currYear] = totalLaborCostPerYear[currYear] ? roundNumber(totalLaborCostPerYear[currYear] + wagesTotalPerYear[currYear], 2) : wagesTotalPerYear[currYear];
      totalVacationHoursGrantedPerYear[currYear] = totalVacationHoursGrantedPerYear[currYear] ? roundNumber(totalVacationHoursGrantedPerYear[currYear] + vacationHoursThisYear, 2) : vacationHoursThisYear;
      totalProductiveHoursPerYear[currYear] = totalProductiveHoursPerYear[currYear] ? roundNumber(totalProductiveHoursPerYear[currYear] + productiveHoursThisYear, 2) : productiveHoursThisYear;
    }

    employeeWageData[employeeId]["totalHourlyWagesPaid"] = wagesTotalPerYear;
    employeeWageData[employeeId]["totalWagesAndBonus"] = wagesAndBonusPaidPerYear;
    employeeWageData[employeeId]["vacationHoursPerYear"] = vacationHoursPerYear;
    employeeWageData[employeeId]["productiveHoursPerYear"] = productiveHoursPerYear;
  }
  const results = { totalLaborCostPerYear: totalLaborCostPerYear, totalVacationHoursGrantedPerYear: totalVacationHoursGrantedPerYear, totalProductiveHoursPerYear: totalProductiveHoursPerYear };
  return results;
};
