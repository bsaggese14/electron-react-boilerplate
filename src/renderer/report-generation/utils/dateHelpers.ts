import $ from "jquery";

// Month and day of week (starting on Sunday) are zero-indexed
// https://codegolf.stackexchange.com/questions/178227/get-the-date-of-the-nth-day-of-week-in-a-given-year-and-month
export const getDateFromData = ({ year, month, dayOfWeek, nthDayOfMonth }: { year: number; month: number; dayOfWeek: number; nthDayOfMonth: number }) => {
  return ((dayOfWeek + 6 - new Date(year, month, 7).getDay()) % 7) + nthDayOfMonth * 7 - 6;
};

const getFirstSundayOfYearDayOfMonth = (year: number) => {
  return getDateFromData({ year, month: 0, dayOfWeek: 0, nthDayOfMonth: 1 });
};

export const getFirstSundayOfYear = (year: number) => {
  return new Date(`01/${getFirstSundayOfYearDayOfMonth(year)}/${year}`);
};

export const workingDaysBetweenDates = ({ startDate, endDate, checkForHolidays = true }: { startDate: Date; endDate: Date; checkForHolidays?: boolean }) => {
  const holidays: Date[] = []; // [new Date("12/25/2023")] // TO DO: sync holidays across the board

  let count = 0;
  const curDate = new Date(startDate.getTime());
  while (curDate <= endDate) {
    const dayOfWeek = curDate.getUTCDay();
    if (dayOfWeek !== 0 && dayOfWeek !== 6) count++;
    curDate.setDate(curDate.getDate() + 1);
  }
  if (checkForHolidays) {
    holidays.forEach((day: Date) => {
      if (day >= startDate && day <= endDate) {
        /* If it is not saturday (6) or sunday (0), substract it */
        if (day.getDay() % 6 != 0) {
          count--;
        }
      }
    });
  }
  return count;
};

export const newDateWithAddedDays = (date: Date, daysToAdd: number) => {
  var newDate = new Date(date);
  newDate.setDate(date.getDate() + daysToAdd);
  return newDate;
};

export const getNextSunday = (startDate: Date) => {
  const weekdayOfStartDate = startDate.getDay();
  const daysToAdd = 7 - weekdayOfStartDate;
  const firstSundayAfterStart = newDateWithAddedDays(startDate, daysToAdd);
  return firstSundayAfterStart;
};

// TO DO: support future
export const getGwiDateFromDom = (currYear: number, reportType: any) => {
  const gwiNth = $(`.${reportType} .gwiNth`).val() as number;
  const gwiDayOfWeek = $(`.${reportType} .gwiDayOfWeek`).val() as number;
  const gwiMonth = $(`.${reportType} .gwiMonth`).val() as number;

  const gwiDateNumber = getDateFromData({ year: currYear, month: gwiMonth - 1, dayOfWeek: gwiDayOfWeek - 1, nthDayOfMonth: gwiNth });
  return `${gwiMonth}/${gwiDateNumber}/${currYear}`;
};

// unsused
const workingDaysBetweenDatesForSchedule = (startDate: any, endDate: any, workdaysPerWeek: number, checkForHolidays = true) => {
  return workingDaysBetweenDates({ startDate, endDate, checkForHolidays }) - 52 * (5 - workdaysPerWeek);
};
