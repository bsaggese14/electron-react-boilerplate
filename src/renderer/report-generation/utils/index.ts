export * from "./parseCurrency";
export * from "./testData";
export * from "./delay";
export * from "./roundNumber";
export * from "./tableHelpers";
export * from "./dateHelpers";
export * from "./report-functions";
