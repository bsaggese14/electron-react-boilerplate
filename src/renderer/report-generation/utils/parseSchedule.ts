// unused
export const hoursePerWeekFromSchedule = (schedule: string) => {
  const splitSchedule = schedule.split("X");
  const daysPerWeek = parseInt(splitSchedule[0]);
  const hoursPerDay = parseInt(splitSchedule[1]);
  const hoursPerWeek = daysPerWeek * hoursPerDay;
  return hoursPerWeek;
};
