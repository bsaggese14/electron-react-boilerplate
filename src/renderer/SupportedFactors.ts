import { GwiFactor } from "./components/factors/gwi/Gwi";
import { VacationFactor } from "./components/factors/vacation/Vacation";

export enum SupportedFactors {
  GWI = "gwi",
  VACATION = "vacation",
}

export type FactorName = {
  name: SupportedFactors;
};

export type Factor = GwiFactor | VacationFactor;
