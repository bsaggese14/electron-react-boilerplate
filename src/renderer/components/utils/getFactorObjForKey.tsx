import { SupportedFactors } from "../../SupportedFactors";
import { GwiFactor } from "../factors/gwi/Gwi";
import { VacationFactor } from "../factors/vacation/Vacation";

const getEmptyVacationObject = () => {
  return { name: SupportedFactors.VACATION, vacationObjects: [] } as VacationFactor;
};

const getEmptyGwiObject = () => {
  return { name: SupportedFactors.GWI, effectiveDate: new Date(), gwiObjects: [] } as GwiFactor;
};

export const getFactorObjForKey = (value: SupportedFactors) => {
  switch (value) {
    case SupportedFactors.GWI: {
      return getEmptyGwiObject();
    }
    case SupportedFactors.VACATION: {
      return getEmptyVacationObject();
    }
    default: {
      return getEmptyGwiObject();
    }
  }
};
