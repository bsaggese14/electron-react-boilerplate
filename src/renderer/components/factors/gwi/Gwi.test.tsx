import { render } from "@testing-library/react";
import { Gwi } from "./Gwi";
import { testGwiObjects } from "../../../report-generation/utils";
import { SupportedFactors } from "../../../SupportedFactors";

const testConfig = { name: SupportedFactors.GWI, effectiveDate: new Date(), gwiObjects: testGwiObjects };
describe("GwiTable", () => {
  it("renders", () => {
    const { container } = render(<Gwi config={testConfig}></Gwi>);
    expect(container).toMatchSnapshot();
  });
});
