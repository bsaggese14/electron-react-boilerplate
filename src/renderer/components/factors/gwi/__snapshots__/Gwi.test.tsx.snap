// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GwiTable renders 1`] = `
<div>
  <h4>
    General Wage Increase (GWI)
  </h4>
  <div>
    <span
      class="me-2"
    >
      Effective
    </span>
    <select
      class="gwiNth"
      name="nth"
    >
      <option
        value="1"
      >
        1st
      </option>
      <option
        value="2"
      >
        2nd
      </option>
      <option
        value="3"
      >
        3rd
      </option>
      <option
        value="4"
      >
        4th
      </option>
    </select>
    <select
      class="gwiDayOfWeek"
      name="dayOfWeek"
    >
      <option
        value="1"
      >
        Monday
      </option>
      <option
        value="2"
      >
        Tuesday
      </option>
      <option
        value="3"
      >
        Wednesday
      </option>
      <option
        value="4"
      >
        Thursday
      </option>
      <option
        value="5"
      >
        Friday
      </option>
      <option
        selected=""
        value="6"
      >
        Saturday
      </option>
      <option
        value="7"
      >
        Sunday
      </option>
    </select>
    <span
      class="mx-1"
    >
      of
    </span>
    <select
      class="gwiMonth"
      name="month"
    >
      <option
        value="1"
      >
        Jan
      </option>
      <option
        value="2"
      >
        Feb
      </option>
      <option
        value="3"
      >
        Mar
      </option>
      <option
        selected=""
        value="4"
      >
        Apr
      </option>
      <option
        value="5"
      >
        May
      </option>
      <option
        value="6"
      >
        Jun
      </option>
      <option
        value="7"
      >
        Jul
      </option>
      <option
        value="8"
      >
        Aug
      </option>
      <option
        value="9"
      >
        Sep
      </option>
      <option
        value="10"
      >
        Oct
      </option>
      <option
        value="11"
      >
        Nov
      </option>
      <option
        value="12"
      >
        Dec
      </option>
    </select>
  </div>
  <table
    class="table table-sm gwiTable"
  >
    <thead>
      <tr>
        <th
          scope="row"
        >
          Year
        </th>
        <th
          scope="col"
        >
          2018
        </th>
        <th
          scope="col"
        >
          2019
        </th>
        <th
          scope="col"
        >
          2020
        </th>
        <th
          scope="col"
        >
          2021
        </th>
        <th
          scope="col"
        >
          2022
        </th>
      </tr>
    </thead>
    <tbody>
      <tr
        class="rowPercent"
      >
        <th
          scope="row"
        >
          Percent
        </th>
        <td>
          <input
            type="text"
            value="3.5"
          />
          %
        </td>
        <td>
          <input
            type="text"
            value="3"
          />
          %
        </td>
        <td>
          <input
            type="text"
            value="3"
          />
          %
        </td>
        <td>
          <input
            type="text"
            value="2.5"
          />
          %
        </td>
        <td>
          <input
            type="text"
            value="2.5"
          />
          %
        </td>
      </tr>
      <tr
        class="rowBonus"
      >
        <th
          scope="row"
        >
          Bonus
        </th>
        <td>
          <input
            type="text"
            value="3500"
          />
        </td>
        <td>
          <input
            type="text"
            value="0"
          />
        </td>
        <td>
          <input
            type="text"
            value="0"
          />
        </td>
        <td>
          <input
            type="text"
            value="0"
          />
        </td>
        <td>
          <input
            type="text"
            value="0"
          />
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
